🤖 backported "feat(embed-js): Add analytics to Embedded Analytics JS"

github-automation-metabase opened the pull request:
#63580

  Conflicts:
  ```shell
  100644 fe03da59c2beb6b091422493efa76499e1e2e208 1	enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute.tsx
100644 8a54138ffc637ce50f2fe58dd04afcaa778905f2 2	enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute.tsx
100644 a66589bce02ff16afef2d373e1e275d0d987eb56 3	enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute.tsx
  ```

  - [x] Conflicts resolved

I tested with this snippet
```html
<metabase-browser initial-collection="root" read-only="true"></metabase-browser>
<metabase-browser initial-collection="root" read-only="false"></metabase-browser>
<metabase-browser initial-collection="root"></metabase-browser>
<metabase-dashboard></metabase-dashboard>
<metabase-question></metabase-question>
<metabase-question question-id="new"></metabase-question>
```

Here's the result snowplow event. I also have retested after chery-picking the last commit https://github.com/metabase/metabase/pull/63741/commits/ace4d9da164f1402525ec5ac568e3f3feb636dba. 

The preview doesn't fire the event as expected, but the actual embed will do.

<details>

<summary>
Click to expand
</summary>

```json
[
  {
    "rawEvent": {
      "api": {
        "vendor": "com.snowplowanalytics.snowplow",
        "version": "tp2"
      },
      "parameters": {
        "e": "ue",
        "eid": "e3b97837-3b6f-479b-bb5d-7cbbfcabc832",
        "url": "http://localhost:3000/embed/sdk/v1",
        "refr": "http://localhost:55187/",
        "aid": "metabase",
        "cx": "eyJzY2hlbWEiOiJpZ2x1OmNvbS5zbm93cGxvd2FuYWx5dGljcy5zbm93cGxvdy9jb250ZXh0cy9qc29uc2NoZW1hLzEtMC0wIiwiZGF0YSI6W3sic2NoZW1hIjoiaWdsdTpjb20ubWV0YWJhc2UvaW5zdGFuY2UvanNvbnNjaGVtYS8xLTEtMCIsImRhdGEiOnsiaWQiOiJjNWE3ZGFhMy0yMGIzLTRmOGYtOGEyMi00ZWEwNmNiNzNhMzMiLCJ2ZXJzaW9uIjp7InRhZyI6InYxLjU3LjEtU05BUFNIT1QifSwiY3JlYXRlZF9hdCI6IjIwMjUtMDktMTlUMDY6MTI6NDkuNjNaIiwidG9rZW5fZmVhdHVyZXMiOnsiZGF0YWJhc2VfYXV0aF9wcm92aWRlcnMiOnRydWUsImNvbmZpZ190ZXh0X2ZpbGUiOnRydWUsInNzb19nb29nbGUiOnRydWUsImFpX3NxbF9nZW5lcmF0aW9uIjp0cnVlLCJzc29fand0Ijp0cnVlLCJsbG1fYXV0b2Rlc2NyaXB0aW9uIjpmYWxzZSwiY2FjaGVfZ3JhbnVsYXJfY29udHJvbHMiOnRydWUsImVtYWlsX3Jlc3RyaWN0X3JlY2lwaWVudHMiOnRydWUsImV0bF9jb25uZWN0aW9uc19wZyI6ZmFsc2UsImFpX2VudGl0eV9hbmFseXNpcyI6dHJ1ZSwiY2xvdWRfY3VzdG9tX3NtdHAiOmZhbHNlLCJ3aGl0ZWxhYmVsIjp0cnVlLCJzYW5kYm94ZXMiOnRydWUsImVtYWlsX2FsbG93X2xpc3QiOnRydWUsInVwbG9hZF9tYW5hZ2VtZW50Ijp0cnVlLCJkYXRhYmFzZV9yb3V0aW5nIjp0cnVlLCJkZXZlbG9wbWVudF9tb2RlIjpmYWxzZSwic2NpbSI6dHJ1ZSwiYXR0YWNoZWRfZHdoIjp0cnVlLCJob3N0aW5nIjpmYWxzZSwic2Vzc2lvbl90aW1lb3V0X2NvbmZpZyI6dHJ1ZSwiY2FjaGVfcHJlZW1wdGl2ZSI6dHJ1ZSwib2ZmZXJfbWV0YWJhc2VfYWkiOmZhbHNlLCJhaV9zcWxfZml4ZXIiOnRydWUsImNvbnRlbnRfdHJhbnNsYXRpb24iOnRydWUsImRpc2FibGVfcGFzc3dvcmRfbG9naW4iOnRydWUsInRhYmxlX2RhdGFfZWRpdGluZyI6dHJ1ZSwib2ZmaWNpYWxfY29sbGVjdGlvbnMiOnRydWUsInNzb19zYW1sIjp0cnVlLCJlbWJlZGRpbmciOnRydWUsImVtYmVkZGluZ19zaW1wbGUiOnRydWUsImV0bF9jb25uZWN0aW9ucyI6dHJ1ZSwiY29sbGVjdGlvbl9jbGVhbnVwIjp0cnVlLCJzc29fbGRhcCI6dHJ1ZSwiY29udGVudF92ZXJpZmljYXRpb24iOnRydWUsImRhc2hib2FyZF9zdWJzY3JpcHRpb25fZmlsdGVycyI6dHJ1ZSwiZW1iZWRkaW5nX3NkayI6dHJ1ZSwiYWR2YW5jZWRfcGVybWlzc2lvbnMiOnRydWUsImF1ZGl0X2FwcCI6dHJ1ZSwic2VyaWFsaXphdGlvbiI6dHJ1ZSwicXVlcnlfcmVmZXJlbmNlX3ZhbGlkYXRpb24iOnRydWUsInNlbWFudGljX3NlYXJjaCI6dHJ1ZSwic25pcHBldF9jb2xsZWN0aW9ucyI6dHJ1ZSwibWV0YWJvdF92MyI6dHJ1ZX19fSx7InNjaGVtYSI6ImlnbHU6Y29tLnNub3dwbG93YW5hbHl0aWNzLnNub3dwbG93L3dlYl9wYWdlL2pzb25zY2hlbWEvMS0wLTAiLCJkYXRhIjp7ImlkIjoiMmUyZTBjNWMtZjZhNS00NTNkLWJlOTEtMzRkOTVhNTQyODIyIn19XX0",
        "tna": "sp",
        "cs": "UTF-8",
        "cd": "24",
        "stm": "1758262765631",
        "tv": "js-3.1.6",
        "vp": "2451x1312",
        "ds": "2451x1312",
        "res": "2560x1440",
        "cookie": "1",
        "p": "web",
        "dtm": "1758262765631",
        "lang": "en-US",
        "ue_px": "eyJzY2hlbWEiOiJpZ2x1OmNvbS5zbm93cGxvd2FuYWx5dGljcy5zbm93cGxvdy91bnN0cnVjdF9ldmVudC9qc29uc2NoZW1hLzEtMC0wIiwiZGF0YSI6eyJzY2hlbWEiOiJpZ2x1OmNvbS5tZXRhYmFzZS9lbWJlZGRlZF9hbmFseXRpY3NfanMvanNvbnNjaGVtYS8xLTAtMCIsImRhdGEiOnsiYnJvd3NlciI6eyJyZWFkX29ubHkiOnsiZmFsc2UiOjEsInRydWUiOjJ9fSwiZXhwbG9yYXRpb24iOnsiaXNfc2F2ZV9lbmFibGVkIjp7ImZhbHNlIjoxLCJ0cnVlIjowfX0sImdsb2JhbCI6eyJhdXRoX21ldGhvZCI6InNlc3Npb24ifSwiZXZlbnQiOiJzZXR1cCIsImRhc2hib2FyZCI6eyJ3aXRoX3RpdGxlIjp7ImZhbHNlIjowLCJ0cnVlIjoxfSwid2l0aF9kb3dubG9hZHMiOnsiZmFsc2UiOjEsInRydWUiOjB9LCJkcmlsbHMiOnsiZmFsc2UiOjAsInRydWUiOjF9fSwicXVlc3Rpb24iOnsiZHJpbGxzIjp7ImZhbHNlIjowLCJ0cnVlIjoxfSwid2l0aF9kb3dubG9hZHMiOnsiZmFsc2UiOjEsInRydWUiOjB9LCJ3aXRoX3RpdGxlIjp7ImZhbHNlIjowLCJ0cnVlIjoxfSwiaXNfc2F2ZV9lbmFibGVkIjp7ImZhbHNlIjoxLCJ0cnVlIjowfX19fX0"
      },
      "contentType": "application/json; charset=utf-8",
      "source": {
        "name": "micro-ssc-2.3.0-memorysink",
        "encoding": "UTF-8",
        "hostname": null
      },
      "context": {
        "timestamp": "2025-09-19T06:19:25.643Z",
        "ipAddress": "unknown",
        "useragent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36",
        "refererUri": "http://localhost:3000/",
        "headers": [
          "Host: localhost:9090",
          "Connection: keep-alive",
          "Content-Length: 3181",
          "sec-ch-ua-platform: \"macOS\"",
          "User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36",
          "SP-Anonymous: *",
          "sec-ch-ua: \"Chromium\";v=\"140\", \"Not=A?Brand\";v=\"24\", \"Google Chrome\";v=\"140\"",
          "Content-Type: application/json; charset=UTF-8",
          "sec-ch-ua-mobile: ?0",
          "Accept: */*",
          "Origin: http://localhost:3000",
          "Sec-Fetch-Site: same-site",
          "Sec-Fetch-Mode: cors",
          "Sec-Fetch-Dest: empty",
          "Referer: http://localhost:3000/",
          "Accept-Encoding: gzip, deflate, br, zstd",
          "Accept-Language: en-US,en;q=0.9",
          "application/json; charset=utf-8"
        ],
        "userId": "00000000-0000-0000-0000-000000000000"
      }
    },
    "eventType": "unstruct",
    "schema": "iglu:com.metabase/embedded_analytics_js/jsonschema/1-0-0",
    "contexts": [
      "iglu:com.metabase/instance/jsonschema/1-1-0",
      "iglu:com.snowplowanalytics.snowplow/web_page/jsonschema/1-0-0"
    ],
    "event": {
      "app_id": "metabase",
      "platform": "web",
      "etl_tstamp": "2025-09-19T06:19:25.646Z",
      "collector_tstamp": "2025-09-19T06:19:25.643Z",
      "dvce_created_tstamp": "2025-09-19T06:19:25.631Z",
      "event": "unstruct",
      "event_id": "e3b97837-3b6f-479b-bb5d-7cbbfcabc832",
      "txn_id": null,
      "name_tracker": "sp",
      "v_tracker": "js-3.1.6",
      "v_collector": "micro-ssc-2.3.0-memorysink",
      "v_etl": "snowplow-micro-2.3.0",
      "user_id": null,
      "user_ipaddress": "unknown",
      "user_fingerprint": null,
      "domain_userid": null,
      "domain_sessionidx": null,
      "network_userid": "00000000-0000-0000-0000-000000000000",
      "geo_country": null,
      "geo_region": null,
      "geo_city": null,
      "geo_zipcode": null,
      "geo_latitude": null,
      "geo_longitude": null,
      "geo_region_name": null,
      "ip_isp": null,
      "ip_organization": null,
      "ip_domain": null,
      "ip_netspeed": null,
      "page_url": "http://localhost:3000/embed/sdk/v1",
      "page_title": null,
      "page_referrer": "http://localhost:55187/",
      "page_urlscheme": "http",
      "page_urlhost": "localhost",
      "page_urlport": 3000,
      "page_urlpath": "/embed/sdk/v1",
      "page_urlquery": null,
      "page_urlfragment": null,
      "refr_urlscheme": "http",
      "refr_urlhost": "localhost",
      "refr_urlport": 55187,
      "refr_urlpath": "/",
      "refr_urlquery": null,
      "refr_urlfragment": null,
      "refr_medium": null,
      "refr_source": null,
      "refr_term": null,
      "mkt_medium": null,
      "mkt_source": null,
      "mkt_term": null,
      "mkt_content": null,
      "mkt_campaign": null,
      "contexts": {
        "schema": "iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0",
        "data": [
          {
            "schema": "iglu:com.metabase/instance/jsonschema/1-1-0",
            "data": {
              "id": "c5a7daa3-20b3-4f8f-8a22-4ea06cb73a33",
              "version": {
                "tag": "v1.57.1-SNAPSHOT"
              },
              "created_at": "2025-09-19T06:12:49.63Z",
              "token_features": {
                "database_auth_providers": true,
                "config_text_file": true,
                "sso_google": true,
                "ai_sql_generation": true,
                "sso_jwt": true,
                "llm_autodescription": false,
                "cache_granular_controls": true,
                "email_restrict_recipients": true,
                "etl_connections_pg": false,
                "ai_entity_analysis": true,
                "cloud_custom_smtp": false,
                "whitelabel": true,
                "sandboxes": true,
                "email_allow_list": true,
                "upload_management": true,
                "database_routing": true,
                "development_mode": false,
                "scim": true,
                "attached_dwh": true,
                "hosting": false,
                "session_timeout_config": true,
                "cache_preemptive": true,
                "offer_metabase_ai": false,
                "ai_sql_fixer": true,
                "content_translation": true,
                "disable_password_login": true,
                "table_data_editing": true,
                "official_collections": true,
                "sso_saml": true,
                "embedding": true,
                "embedding_simple": true,
                "etl_connections": true,
                "collection_cleanup": true,
                "sso_ldap": true,
                "content_verification": true,
                "dashboard_subscription_filters": true,
                "embedding_sdk": true,
                "advanced_permissions": true,
                "audit_app": true,
                "serialization": true,
                "query_reference_validation": true,
                "semantic_search": true,
                "snippet_collections": true,
                "metabot_v3": true
              }
            }
          },
          {
            "schema": "iglu:com.snowplowanalytics.snowplow/web_page/jsonschema/1-0-0",
            "data": {
              "id": "2e2e0c5c-f6a5-453d-be91-34d95a542822"
            }
          }
        ]
      },
      "se_category": null,
      "se_action": null,
      "se_label": null,
      "se_property": null,
      "se_value": null,
      "unstruct_event": {
        "schema": "iglu:com.snowplowanalytics.snowplow/unstruct_event/jsonschema/1-0-0",
        "data": {
          "schema": "iglu:com.metabase/embedded_analytics_js/jsonschema/1-0-0",
          "data": {
            "browser": {
              "read_only": {
                "false": 1,
                "true": 2
              }
            },
            "exploration": {
              "is_save_enabled": {
                "false": 1,
                "true": 0
              }
            },
            "global": {
              "auth_method": "session"
            },
            "event": "setup",
            "dashboard": {
              "with_title": {
                "false": 0,
                "true": 1
              },
              "with_downloads": {
                "false": 1,
                "true": 0
              },
              "drills": {
                "false": 0,
                "true": 1
              }
            },
            "question": {
              "drills": {
                "false": 0,
                "true": 1
              },
              "with_downloads": {
                "false": 1,
                "true": 0
              },
              "with_title": {
                "false": 0,
                "true": 1
              },
              "is_save_enabled": {
                "false": 1,
                "true": 0
              }
            }
          }
        }
      },
      "tr_orderid": null,
      "tr_affiliation": null,
      "tr_total": null,
      "tr_tax": null,
      "tr_shipping": null,
      "tr_city": null,
      "tr_state": null,
      "tr_country": null,
      "ti_orderid": null,
      "ti_sku": null,
      "ti_name": null,
      "ti_category": null,
      "ti_price": null,
      "ti_quantity": null,
      "pp_xoffset_min": null,
      "pp_xoffset_max": null,
      "pp_yoffset_min": null,
      "pp_yoffset_max": null,
      "useragent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36",
      "br_name": null,
      "br_family": null,
      "br_version": null,
      "br_type": null,
      "br_renderengine": null,
      "br_lang": "en-US",
      "br_features_pdf": null,
      "br_features_flash": null,
      "br_features_java": null,
      "br_features_director": null,
      "br_features_quicktime": null,
      "br_features_realplayer": null,
      "br_features_windowsmedia": null,
      "br_features_gears": null,
      "br_features_silverlight": null,
      "br_cookies": true,
      "br_colordepth": "24",
      "br_viewwidth": 2451,
      "br_viewheight": 1312,
      "os_name": null,
      "os_family": null,
      "os_manufacturer": null,
      "os_timezone": null,
      "dvce_type": null,
      "dvce_ismobile": null,
      "dvce_screenwidth": 2560,
      "dvce_screenheight": 1440,
      "doc_charset": "UTF-8",
      "doc_width": 2451,
      "doc_height": 1312,
      "tr_currency": null,
      "tr_total_base": null,
      "tr_tax_base": null,
      "tr_shipping_base": null,
      "ti_currency": null,
      "ti_price_base": null,
      "base_currency": null,
      "geo_timezone": null,
      "mkt_clickid": null,
      "mkt_network": null,
      "etl_tags": null,
      "dvce_sent_tstamp": "2025-09-19T06:19:25.631Z",
      "refr_domain_userid": null,
      "refr_dvce_tstamp": null,
      "derived_contexts": {

      },
      "domain_sessionid": null,
      "derived_tstamp": "2025-09-19T06:19:25.643Z",
      "event_vendor": "com.metabase",
      "event_name": "embedded_analytics_js",
      "event_format": "jsonschema",
      "event_version": "1-0-0",
      "event_fingerprint": null,
      "true_tstamp": null
    }
  }
]
```

</details>

WiNloSt committed:
feat(embed-js): Add analytics to Embedded Analytics JS (#63580)

* Add Embedded Analytics JS usage snowplow schema

* Add analytics

* Clarify code with comments

* Fix analytics logic + fix typing

* Add unit tests

* Make E2E embedded analytics JS util supports multiple elements

* Fix code not working in Cypress for some reason

* Add E2E test suite

* Fix import order

* Remove extra line

* Fix E2E from util API change

* Update snowplow/iglu-client-embedded/schemas/com.metabase/embedded_analytics_js/jsonschema/1-0-0

Co-authored-by: Phoomparin Mano <poom@metabase.com>

* Update e2e/test/scenarios/embedding/sdk-iframe-embedding/sdk-iframe-embedding.cy.spec.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

* Update e2e/support/helpers/e2e-embedding-iframe-sdk-helpers.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

* Fix E2E by updating a function API

* Pass snowplow schema lint

---------

Co-authored-by: Phoomparin Mano <poom@metabase.com>
Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

WiNloSt committed:
Use the `reduxStore` prop passed

heypoom reviewed (approved):
Reviewed Kelvin's comments on conflict areas, and the resolution looks good!

WiNloSt committed:
fix(embed-js): Only report analytics when not in the preview (#63812)

* Only report analytics when not in the preview

This would avoid confusing preview usage with the actual usage

* Ensure analytics would be report even if iframe isn't ready

* Add E2E test

* Make the test pass

* Stop relying on redux store to fetch `site-url`

* Clarify test with comment

* Add tests

* Fix linter

I was naming a file wrong, so it doesn't match unit test spec. This rule
apparently doesn't take tests into account, which is great.
